File: task3/c++/Subtask-4.cpp
#include <fstream>
#include <iostream>

void generateDiamond(int n, std::ofstream &outputFile) {
    // Print the upper part of the diamond
    for (int i = 0; i < n; ++i) {
        // Print leading spaces
        for (int j = 0; j < n - i - 1; ++j) {
            outputFile << ' ';
        }
        // Print asterisks
        for (int j = 0; j < 2 * i + 1; ++j) {
            outputFile << '*';
        }
        // Move to the next line
        outputFile << std::endl;
    }

    // Print the lower part of the diamond
    for (int i = n - 2; i >= 0; --i) {
        // Print leading spaces
        for (int j = 0; j < n - i - 1; ++j) {
            outputFile << ' ';
        }
        // Print asterisks
        for (int j = 0; j < 2 * i + 1; ++j) {
            outputFile << '*';
        }
        // Move to the next line
        outputFile << std::endl;
    }
}

int main() {
    int n;
    std::ifstream inputFile("vsofficial.txt"); // Open input file
    std::ofstream outputFile("output.txt");    // Open output file

    if (!inputFile.is_open()) {
        std::cerr << "Error opening input file." << std::endl;
        return 1;
    }
    if (!outputFile.is_open()) {
        std::cerr << "Error opening output file." << std::endl;
        return 1;
    }

    // Read the number from the input file
    inputFile >> n;
    inputFile.close();

    // Validate input
    if (n <= 0) {
        std::cerr << "Number of rows must be a positive integer." << std::endl;
        return 1;
    }

    // Generate and write the diamond pattern to the output file
    generateDiamond(n, outputFile);

    outputFile.close();

    std::cout << "Diamond pattern written to output.txt" << std::endl;

    return 0;
}

File: task3/Rust/Subtask4.rs
use std::fs::File;
use std::io::{self, Read, Write};

fn main() -> io::Result<()> {
    // Open the input file vsofficial.txt
    let mut input_file = File::open("vsofficial.txt")?;
    
    // Create a buffer to store the number
    let mut content = String::new();
    
    // Read the content of the input file
    input_file.read_to_string(&mut content)?;
    
    // Parse the content into an integer
    let n: usize = content.trim().parse().expect("Please enter a valid number in vsofficial.txt");
    
    // Open (or create) the output file output.txt
    let mut output_file = File::create("output.txt")?;
    
    // Generate the top half of the diamond (including the middle row)
    for i in 0..n {
        // Write leading spaces to the file
        for _ in 0..(n - i - 1) {
            write!(output_file, " ")?;
        }
        // Write asterisks to the file
        for _ in 0..(2 * i + 1) {
            write!(output_file, "*")?;
        }
        // Move to the next line in the file
        writeln!(output_file)?;
    }
    
    // Generate the bottom half of the diamond
    for i in (0..(n - 1)).rev() {
        // Write leading spaces to the file
        for _ in 0..(n - i - 1) {
            write!(output_file, " ")?;
        }
        // Write asterisks to the file
        for _ in 0..(2 * i + 1) {
            write!(output_file, "*")?;
        }
        // Move to the next line in the file
        writeln!(output_file)?;
    }
    
    println!("Diamond pattern generated in output.txt");
    
    Ok(())
}
